File: backend/app/schemas/address.py
"""
PRD 模块四 · 用户常用地址簿 schemas。

`full_text` 是地址簿与订单解耦的"瞬时快照"字段:
下单时直接写到 orders.shipping_address_json 后,用户之后修改/删除地址簿条目
不会影响已下单的订单。
"""
from typing import Optional
from datetime import datetime
from pydantic import BaseModel, Field


class UserAddress(BaseModel):
    id: int
    receiverName: str
    phone: str
    country: Optional[str] = None
    province: Optional[str] = None
    city: Optional[str] = None
    district: Optional[str] = None
    detail: Optional[str] = None
    fullText: str
    postalCode: Optional[str] = None
    label: Optional[str] = None
    isDefault: bool = False
    createdAt: Optional[datetime] = None
    updatedAt: Optional[datetime] = None


class UserAddressCreate(BaseModel):
    receiverName: str = Field(..., min_length=1, max_length=40)
    phone: str = Field(..., min_length=5, max_length=24)
    country: Optional[str] = Field(None, max_length=40)
    province: Optional[str] = Field(None, max_length=40)
    city: Optional[str] = Field(None, max_length=40)
    district: Optional[str] = Field(None, max_length=40)
    detail: Optional[str] = Field(None, max_length=200)
    fullText: Optional[str] = Field(
        None,
        max_length=400,
        description="地址纯文本。前端可拼接 province+city+district+detail 后传入,后端缺省时按此规则拼。",
    )
    postalCode: Optional[str] = Field(None, max_length=20)
    label: Optional[str] = Field(None, max_length=20)
    isDefault: bool = False


class UserAddressUpdate(BaseModel):
    receiverName: Optional[str] = Field(None, min_length=1, max_length=40)
    phone: Optional[str] = Field(None, min_length=5, max_length=24)
    country: Optional[str] = Field(None, max_length=40)
    province: Optional[str] = Field(None, max_length=40)
    city: Optional[str] = Field(None, max_length=40)
    district: Optional[str] = Field(None, max_length=40)
    detail: Optional[str] = Field(None, max_length=200)
    fullText: Optional[str] = Field(None, max_length=400)
    postalCode: Optional[str] = Field(None, max_length=20)
    label: Optional[str] = Field(None, max_length=20)
    isDefault: Optional[bool] = None

File: web/src/lib/services/address.ts
/**
 * Web 端收货地址簿 API 客户端。
 *
 * 对齐移动端 `frontend/src/services/addressService.ts`，后端入口 `/api/me/addresses`。
 *
 * 与订单的关系：结算页通过地址选择器挑一条地址，把快照传给 buyNow /
 * setShippingAddress。订单上的 shippingAddress 是「下单瞬间快照」，
 * 与地址簿条目解耦——事后改地址簿不会影响已下单的订单。
 */

import { apiClient } from "../api-client";

export interface UserAddress {
  id: number;
  receiverName: string;
  phone: string;
  country?: string | null;
  province?: string | null;
  city?: string | null;
  district?: string | null;
  detail?: string | null;
  fullText: string;
  postalCode?: string | null;
  label?: string | null;
  isDefault: boolean;
  createdAt?: string | null;
  updatedAt?: string | null;
}

export interface UserAddressCreateParams {
  receiverName: string;
  phone: string;
  country?: string | null;
  province?: string | null;
  city?: string | null;
  district?: string | null;
  detail?: string | null;
  fullText?: string | null;
  postalCode?: string | null;
  label?: string | null;
  isDefault?: boolean;
}

export type UserAddressUpdateParams = Partial<UserAddressCreateParams>;

export const addressService = {
  listMyAddresses: async (): Promise<UserAddress[]> => {
    const res = await apiClient.get<{ items: UserAddress[] }>(
      "/api/me/addresses",
    );
    return res?.items ?? [];
  },

  getDefaultAddress: async (): Promise<UserAddress | null> => {
    const res = await apiClient.get<UserAddress | null>(
      "/api/me/addresses/default",
    );
    return res ?? null;
  },

  createAddress: (payload: UserAddressCreateParams) =>
    apiClient.post<UserAddress>("/api/me/addresses", payload),

  updateAddress: (id: number, payload: UserAddressUpdateParams) =>
    apiClient.put<UserAddress>(`/api/me/addresses/${id}`, payload),

  setDefaultAddress: (id: number) =>
    apiClient.post<UserAddress>(`/api/me/addresses/${id}/default`),

  deleteAddress: (id: number) =>
    apiClient.delete<void>(`/api/me/addresses/${id}`),
};

/**
 * 把地址簿条目转成订单需要的收货地址快照。
 * 后端只认这几个字段，多余字段会被忽略。
 */
export function toShippingAddressSnapshot(
  address: UserAddress,
): Record<string, unknown> {
  return {
    receiverName: address.receiverName,
    phone: address.phone,
    country: address.country ?? null,
    province: address.province ?? null,
    city: address.city ?? null,
    district: address.district ?? null,
    detail: address.detail ?? null,
    fullText: address.fullText,
    postalCode: address.postalCode ?? null,
  };
}

/** 从订单的 shippingAddress JSON 里安全地读一个字符串字段。 */
export function readAddressField(
  address: Record<string, unknown> | null | undefined,
  key: string,
): string {
  if (!address) return "";
  const value = address[key];
  return typeof value === "string" ? value : "";
}

/** 订单收货地址的单行展示文本。 */
export function formatShippingAddress(
  address: Record<string, unknown> | null | undefined,
): string {
  if (!address) return "";
  const full = readAddressField(address, "fullText");
  if (full) return full;
  return ["country", "province", "city", "district", "detail"]
    .map((key) => readAddressField(address, key))
    .filter(Boolean)
    .join(" ");
}
